libwallet/assets/dcr/dex_asset.go

```diff
@@ -0,0 +1,511 @@
+func (da *DEXAsset) NotifyOnTipChange(ctx context.Context, cb dexdcr.TipChangeCallback) bool {
+	const dexAssetOnTipCallbackIdentifier = "dex-asset-on-tip-change"
+	listener := &wallet.TxAndBlockNotificationListener{
+		OnBlockAttached: func(walletID int, blockHeight int32) {
+			cb(ctx, nil, int64(blockHeight), nil)
+		},
+	}
+
+	da.asset.RemoveTxAndBlockNotificationListener(dexAssetOnTipCallbackIdentifier) // Clear previous listener if any.
+	err := da.asset.AddTxAndBlockNotificationListener(listener, dexAssetOnTipCallbackIdentifier)
+	return err != nil // TODO
+}
+
+// AddressInfo returns information for the provided address. It is an error
+// if the address is not owned by the wallet.
+// Part of the Wallet interface.
+func (da *DEXAsset) AddressInfo(ctx context.Context, address string) (*dexdcr.AddressInfo, error) {
+	addr, err := stdaddr.DecodeAddress(address, da.asset.chainParams)
+	if err != nil {
+		return nil, err
+	}
+	ka, err := da.asset.Internal().DCR.KnownAddress(ctx, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	if ka, ok := ka.(dcrwallet.BIP0044Address); ok {
+		_, branch, _ := ka.Path()
+		return &dexdcr.AddressInfo{Account: ka.AccountName(), Branch: branch}, nil
+	}
+	return nil, fmt.Errorf("unsupported address type %T", ka)
+}
+
+// AccountOwnsAddress checks if the provided address belongs to the
+// specified account.
+// Part of the Wallet interface.
+func (da *DEXAsset) AccountOwnsAddress(ctx context.Context, addr stdaddr.Address, acctName string) (bool, error) {
+	a, err := da.asset.Internal().DCR.KnownAddress(ctx, addr)
+	if err != nil {
+		return false, utils.TranslateError(err)
+	}
+
+	if a.AccountName() != acctName {
+		return false, nil
+	}
+
+	if kind := a.AccountKind(); kind != dcrwallet.AccountKindBIP0044 && kind != dcrwallet.AccountKindImported {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// AccountBalance returns the balance breakdown for the specified account.
+// Part of the Wallet interface.
+func (da *DEXAsset) AccountBalance(_ context.Context, confirms int32, _ string) (*walletjson.GetAccountBalanceResult, error) {
+	bal, err := da.asset.GetAccountBalance(int32(da.dexAccountNumber), &confirms)
+	if err != nil {
+		return nil, err
+	}
+
+	return &walletjson.GetAccountBalanceResult{
+		AccountName:             da.dexAccountName,
+		ImmatureCoinbaseRewards: bal.ImmatureReward.ToCoin(),
+		ImmatureStakeGeneration: bal.ImmatureStakeGeneration.ToCoin(),
+		LockedByTickets:         bal.LockedByTickets.ToCoin(),
+		Spendable:               bal.Spendable.ToCoin(),
+		Total:                   bal.Total.ToCoin(),
+		Unconfirmed:             bal.UnConfirmed.ToCoin(),
+		VotingAuthority:         bal.VotingAuthority.ToCoin(),
+	}, nil
+}
+
+// LockedOutputs fetches locked outputs for the Wallet.
+// Part of the Wallet interface.
+func (da *DEXAsset) LockedOutputs(ctx context.Context, _ string) ([]chainjson.TransactionInput, error) {
+	return da.asset.Internal().DCR.LockedOutpoints(ctx, da.dexAccountName)
+}
```

Comment:
When mixing is enabled, multiple accounts may be in use. Always use the acctName parameter passed to methods.